sanic/response/types.py

```diff
@@ -0,0 +1,318 @@
+            raise SanicException(
+                "No stream is connected to the response object instance."
+            )
+        if self.stream.send is None:
+            if end_stream and not data:
+                return
+            raise ServerError(
+                "Response stream was ended, no more response data is "
+                "allowed to be sent."
+            )
+        data = (
+            data.encode()  # type: ignore
+            if hasattr(data, "encode")
+            else data or b""
+        )
+        await self.stream.send(
+            data,  # type: ignore
+            end_stream=end_stream or False,
+        )
+
+
+class HTTPResponse(BaseHTTPResponse):
+    """
+    HTTP response to be sent back to the client.
+
+    :param body: the body content to be returned
+    :type body: Optional[bytes]
+    :param status: HTTP response number. **Default=200**
+    :type status: int
+    :param headers: headers to be returned
+    :type headers: Optional;
+    :param content_type: content type to be returned (as a header)
+    :type content_type: Optional[str]
+    """
+
+    __slots__ = ()
+
+    def __init__(
+        self,
+        body: Optional[Any] = None,
+        status: int = 200,
+        headers: Optional[Union[Header, Dict[str, str]]] = None,
+        content_type: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.content_type: Optional[str] = content_type
+        self.body = self._encode_body(body)
+        self.status = status
+        self.headers = Header(headers or {})
+        self._cookies = None
+
+    async def eof(self):
+        await self.send("", True)
+
+    async def __aenter__(self):
+        return self.send
+
+    async def __aexit__(self, *_):
+        await self.eof()
+
+
+class JSONResponse(HTTPResponse):
+    __slots__ = (
+        "_dumps_method",
+        "_dumps_kwargs",
+        "_raw_body",
+    )
+
+    def __init__(
+        self,
+        body: Optional[Any] = None,
+        status: int = 200,
+        headers: Optional[Union[Header, Dict[str, str]]] = None,
+        content_type: Optional[str] = None,
+        dumps: Optional[Callable[..., str]] = None,
+        **kwargs: Any,
+    ):
+        if not dumps:
+            dumps = HTTPResponse._dumps
+        self._dumps_method = dumps
+        self._dumps_kwargs = kwargs
+        self._raw_body = body
+        super().__init__(
+            self._dumps_method(body, **kwargs),
+            headers=headers,
+            status=status,
+            content_type=content_type,
+        )
+
+    @property
+    def raw_body(self) -> Optional[Any]:
+        return self._raw_body
+
+    @raw_body.setter
+    def raw_body(self, value: Any):
+        self.set_json(value)
+
+    def set_json(
+        self,
+        new_json: Any,
+        dumps: Optional[Callable[..., str]] = None,
+        **kwargs: Any,
+    ):
+        dumps_ = dumps or self._dumps_method
+        kwargs_ = kwargs if kwargs else self._dumps_kwargs
```

Comment:
I wonder if we should expose `_dumps_kwargs` as a property?

```python
response = json(...)
kwargs = {**response.dumps_kwargs, ...}
response.set_json(..., **kwargs)
```

This would make it possible to extend more easily. Thoughts?

Reply:
Agreed! Do you think they should be overwritten if new ones are passed when calling `set_json`?

Reply:
Probably not. That would be an odd side effect.